Proposal: support for logging through the log crate

xanathar opened the pull request:
This PR is (as now) more meant as a base for discussion rather than something meant to be merged (though, in case, it's tested and somewhat ready).

This is a follow-up to this issue: https://github.com/gtk-rs/glib/issues/658

The changes contain:
- a `GlibLogger` struct implementing the `log::Log` trait, so that code using the [log crate](https://crates.io/crates/log) will log to glib automatically (with customizable behavior)
- a series of optional macros which override the macros defined in the [log crate](https://crates.io/crates/log), which allow to specify the logging domain in a `G_LOG_DOMAIN` static/const

The two parts above are "activated" by two different features in `Cargo.toml`: `bridged_logging` and `bridged_logging_domain_macros`.

I understand these are probably a bit "opinionated" in their implementation, so feedback is welcome; probably the most controversial part could be the overridden macros: if those are an obstacle they can be removed and only the adapter kept.

I think a sizable part of people interfacing glib code with Rust code using gtk-rs will want to have the two logging systems cooperating; currently this is done by custom code or external third party crates, so this might be helpful.

Example code using this, would be:

```rust
static glib_logger: glib::log_bridge::GlibLogger = glib::log_bridge::GlibLogger {
    format: glib::log_bridge::GlibLoggerFormat::Structured,
    domain: glib::log_bridge::GlibLoggerDomain::CrateTarget,
};

fn main() {
    log::set_logger(&glib_logger);
    log::set_max_level(log::LevelFilter::Debug);

    info!("This line will get logged by glib");

    error!("This is a Rust error");

    info!("This is from Rust %s and doesn't crash");

    g_info!("domain", "This is using glib logging");
}
```

sdroege commented:
>     * a `GlibLogger` struct implementing the `log::Log` trait, so that code using the [log crate](https://crates.io/crates/log) will log to glib automatically (with customizable behavior)

I like this in principle, will review tomorrow :) I wanted to have this in glib since a while.

>     * a series of optional macros which override the macros defined in the [log crate](https://crates.io/crates/log), which allow to specify the logging domain in a `G_LOG_DOMAIN` static/const

This is confusing me a bit. How does that differ from [`g_info!`](https://gtk-rs.org/docs/glib/macro.g_info.html) etc?

xanathar commented:
> This is confusing me a bit. How does that differ from g_info! etc?

Two ways mainly:
- they are named like the log crate macros (so they are quite a "plug&play" replacement, changing the `macro_use` in `lib.rs`/`main.rs` is usually enough)
- they default to use a variable named `G_LOG_DOMAIN ` (which must be in scope) if no context is provided, thus sort of emulating what `G_LOG_DOMAIN ` does in C.

As I said, those are the most opinionated part.

sdroege commented on src/bridged_logging.rs:
```diff
@@ -0,0 +1,226 @@
+        };
+
+        unsafe {
+            crate::glib_sys::g_log_structured_standard(
```
The structured logging feature is new in GLib 2.50, but as we don't actually log any custom structured information just going via `g_log` should be sufficient or not?

sdroege replied:
Oh nevermind, actually we do. And this function is new in 2.56 :) So needs some condition compilation around the `v2_56` feature.

sdroege commented on src/bridged_logging.rs:
```diff
@@ -0,0 +1,226 @@
+        log::log!(target: $target, log::Level::Trace, $($arg)+);
+    );
+    ($($arg:tt)+) => (
+        log::log!(target: G_LOG_DOMAIN, log::Level::Trace, $($arg)+);
```
I didn't know that the macro hygiene rules allowed this, interesting.

But can you put these macros into a separate commit in the same PR?

xanathar replied:
Sure, I'll do one forced commit without and another with.

sdroege commented on src/lib.rs:
```diff
@@ -198,6 +199,9 @@ pub use log::{
 };
 
+#[cfg(any(feature = "dox", all(feature = "v2_46", feature = "bridged_logging")))]
+pub mod bridged_logging;
```
Maybe `pub use bridged_logging::GlibLogger`?

sdroege commented on src/bridged_logging.rs:
````diff
@@ -0,0 +1,226 @@
+/// ```
+pub struct GlibLogger {
+    pub format: GlibLoggerFormat,
+    pub domain: GlibLoggerDomain,
````
Maybe make these private and private a constructor

sdroege commented on src/bridged_logging.rs:
```diff
@@ -0,0 +1,226 @@
+    /// Logs will use the specified domain for logging. Note that this option
+    /// due to the way the `log` crate is implemented, is global for all crates
+    /// in the current process.
+    Custom(&'static str),
```
Is this really useful?

xanathar replied:
I'm on the fence here.
On one side I assume there will be somebody who wants to add a Rust library to an existing glib based project and is happy to specify a single domain for all Rust logging; on the other they can get a better more future-proof result with a little more effort specifying a domain everywhere.

sdroege replied:
I don't know, @GuillaumeGomez ? :)

GuillaumeGomez replied:
I think it's a good idea. However, a static `str` might be too restrictive.

sdroege replied:
Must be either that or a `String`

GuillaumeGomez replied:
A `string` would be preferable indeed.

xanathar replied:
I fear a `String` cannot be done (at the very least not without some dirty tricks), because the logger is a `&'static` and, as far as I can tell, there is no way to build a `static String` (all its constructors are not `const fn`).

Possible workarounds are with a `lazy_static` or with a `RefCell` inside the logger, but at that point I fear it becomes dirtier and harder to use correctly vs the potential gains.

*(disclaimer: limited knowledge of Rust here, so I might be missing something)

sdroege replied:
You could use [`set_boxed_logger()`](https://docs.rs/log/0.4.11/log/fn.set_boxed_logger.html) but I agree that it's a bit ugly.

So what would be the use-case for this custom string, how would one use it and isn't it confusing to just have everything going into the `log` crate showing up the same way in the glib logs?

xanathar replied:
My 2 cents given the current state: let's remove this option, move forward with the rest and eventually add it in the future in another pr if and when any reasonable need+solution comes up.

sdroege replied:
Sounds good to me :)

GuillaumeGomez replied:
Sounds good to me as well.

sdroege commented:
Looks good to me generally

xanathar commented:
Sorry I mistakenly closed the PR.

xanathar commented:
CI failures seem to be related to formatting. I will fix them by today, along with a couple of docs comments which are not accurate after the latest changes, and I will undraft this.

Open questions:
- What do we want to do with `GlibLoggerDomain::Custom(&'static str)`
- What to do with the override macros (is a separate commit enough)

sdroege commented:
> What do we want to do with `GlibLoggerDomain::Custom(&'static str)`

I don't see its point personally, but maybe someone else does? :)

> What to do with the override macros (is a separate commit enough)

I like them and just want a second opinion.

GuillaumeGomez commented:
I like the macros too. :)

sdroege commented:
It looks like the macros disappeared from the changes somehow?

sdroege commented on src/bridged_logging.rs:
```diff
@@ -0,0 +1,175 @@
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+#![cfg(any(feature = "dox", all(feature = "v2_46", feature = "bridged_logging")))]
```
We usually put those above the `mod` keyword

sdroege commented on Cargo.toml:
```diff
@@ -48,7 +49,8 @@ v2_58 = ["v2_56", "glib-sys/v2_58", "gobject-sys/v2_58"]
 v2_62 = ["v2_60", "glib-sys/v2_62", "gobject-sys/v2_62"]
 v2_64 = ["v2_62", "glib-sys/v2_64"]
-dox = ["glib-sys/dox", "gobject-sys/dox"]
+bridged_logging = ["v2_46", "log"]
```
Which part needs GLib 2.46?

xanathar replied:
Ooops. My eyes had been tricked by 

```rust
mod log;
#[cfg(any(feature = "v2_46", feature = "dox"))]
pub use log::log_set_handler;
```

You are right, there's no need for v2_46 here.

sdroege commented on src/bridged_logging.rs:
```diff
@@ -0,0 +1,175 @@
+}
+
+/// Enumeration of the possible domain handling behaviours for a [`GlibLogger`]
+pub enum GlibLoggerDomain {
```
Derive Debug, Clone, Copy, PartialEq, Eq :)

sdroege commented on src/bridged_logging.rs:
````diff
@@ -0,0 +1,175 @@
+///     info!("This line will get logged by glib");
+/// }
+/// ```
+pub struct GlibLogger {
````
Derive Debug at least

sdroege commented:
Looks good to me otherwise

xanathar commented:
> It looks like the macros disappeared from the changes somehow?

You told me to do a second separate commit... so I removed them and adding them later in a new commit. Have I misunderstood what you wanted ?

sdroege commented:
> You told me to do a second separate commit... so I removed them and adding them later in a new commit. Have I misunderstood what you wanted ?

A separate commit in this PR. It's a separate functional change, so should be a separate commit but it can be part of the same PR as it's related.

Personally I would also clean up the commit history of this PR so it doesn't contain all the back and forth but just a clean history but I guess we can forget that for the gtk repository anyway :)

xanathar commented:
> Personally I would also clean up the commit history of this PR so it doesn't contain all the back and forth but just a clean history but I guess we can forget that for the gtk repository anyway :)

My github-fu is a bit lacking, but I might have done it without accidentally closing the PR like I did last time 😅

xanathar commented:
Thanks to you :)

GuillaumeGomez commented on Cargo.toml:
```diff
@@ -48,7 +49,9 @@ v2_58 = ["v2_56", "glib-sys/v2_58", "gobject-sys/v2_58"]
 v2_64 = ["v2_62", "glib-sys/v2_64"]
-dox = ["glib-sys/dox", "gobject-sys/dox"]
+bridged_logging = ["log"]
+bridged_logging_domain_macros = ["bridged_logging", "log"]
```
The features' name aren't great... :-/

sdroege replied:
Indeed but I don't have a better suggestion. Do you? :)

GuillaumeGomez replied:
No indeed. Doesn't mean we shouldn't try to find better ones though. :p

sdroege replied:
Other crates seem to just call the feature `log`, so we could do the same. `log` and `log_macros`

GuillaumeGomez replied:
Sounds good to me!

GuillaumeGomez commented on Cargo.toml:
```diff
@@ -48,7 +49,9 @@ v2_58 = ["v2_56", "glib-sys/v2_58", "gobject-sys/v2_58"]
-dox = ["glib-sys/dox", "gobject-sys/dox"]
+bridged_logging = ["log"]
+bridged_logging_domain_macros = ["bridged_logging", "log"]
+dox = ["glib-sys/dox", "gobject-sys/dox", "log"]
```
```suggestion
dox = ["glib-sys/dox", "gobject-sys/dox", "bridged_logging_domain_macros"]
```

GuillaumeGomez commented on src/bridged_logging.rs:
```diff
@@ -0,0 +1,245 @@
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+extern crate log as rs_log;
```
It's weird to import a crate outside of the `lib.rs` file. What do you think @sdroege ?

sdroege replied:
I don't see the problem :)

GuillaumeGomez commented on src/bridged_logging.rs:
```diff
@@ -0,0 +1,245 @@
+///     glib::GlibLoggerDomain::CrateTarget,
+/// );
+///
+/// fn main() {
```
Just a nit but you don't need to put this code in the `main` function.

GuillaumeGomez commented on src/bridged_logging.rs:
```diff
@@ -0,0 +1,245 @@
+    ///     glib::GlibLoggerDomain::CrateTarget,
+    /// );
+    ///
+    /// fn main() {
```
nit: no need for `main` function or any function.

GuillaumeGomez commented on src/bridged_logging.rs:
```diff
@@ -0,0 +1,245 @@
+        }
+    }
+
+    fn write_log(domain: Option<&str>, level: rs_log::Level, message: &str) {
```
Just wondering: isn't this function supposed to be public?

sdroege replied:
No, it's an internal function called from the public one.

GuillaumeGomez commented on src/bridged_logging.rs:
```diff
@@ -0,0 +1,245 @@
+    }
+
+    #[cfg(any(feature = "v2_56", feature = "dox"))]
+    fn write_log_structured(
```
Just wondering: isn't this function supposed to be public?

sdroege replied:
No, it's an internal function called from the public one.

GuillaumeGomez commented on Cargo.toml:
```diff
@@ -48,7 +49,9 @@ v2_58 = ["v2_56", "glib-sys/v2_58", "gobject-sys/v2_58"]
 v2_64 = ["v2_62", "glib-sys/v2_64"]
-dox = ["glib-sys/dox", "gobject-sys/dox"]
+bridged_logging = ["log"]
+bridged_logging_domain_macros = ["bridged_logging", "log"]
```
```suggestion
bridged_logging_domain_macros = ["bridged_logging"]
```

xanathar commented:
Changed the code with all the suggestions collected so far.

sdroege commented on src/lib.rs:
```diff
@@ -198,6 +199,12 @@ pub use log::{
+#[cfg(any(feature = "log", feature = "dox"))]
+#[macro_use]
+mod bridged_logging;
+#[cfg(any(feature = "log", feature = "dox"))]
```
Does this work without declaring a feature for this in `Cargo.toml`? The optional dependency implicitly declares a feature?

xanathar replied:
From what I could get: dependencies and features share the same namespace in Cargo.toml, so any optional dependency is implicitly also something you can specify as a feature.

If you are worried about discoverability, I think we can rename the `log` dependency giving it an alias, and then have a separate feature with the `log` name.

sdroege replied:
As long as `cargo build --features log` works, this is ok for me. Discoverability of features is a whole other story and also suboptimal for other features already. For now this just has to be clear from the documentation.

xanathar replied:
I added some notes in the documentation, now it should be clearer.

sdroege commented on Cargo.toml:
```diff
@@ -48,7 +49,8 @@ v2_58 = ["v2_56", "glib-sys/v2_58", "gobject-sys/v2_58"]
 v2_62 = ["v2_60", "glib-sys/v2_62", "gobject-sys/v2_62"]
 v2_64 = ["v2_62", "glib-sys/v2_64"]
-dox = ["glib-sys/dox", "gobject-sys/dox"]
+log_macros = ["log"]
```
Can you change `.travis.yml` to to also enable both new features in the cases it builds for the latest GLib version currently?

xanathar replied:
I can do it, but looking at `.travis.yml` it seems that it does two builds, one against the empty feature set, and the other against `v2_48`.

Do you want me to add `log_macros` beside `v2_48` or upgrade to the latest, while we are at it (considering that structured logging requires `v2_56`)?

sdroege replied:
Add it besides the other one. I don't think we can update to latest because travis runs an ancient version of Ubuntu, but maybe that changed in the meantime.

xanathar commented:
I've clarified the usage in documentation, fixed the code examples and added the new features to Travis' CI (hoping it gets green :) ).
This should address all feedback received so far.

sdroege commented:
CI seems happy :) Thanks!
